---
title: "10X QC processing"
author: "Kimberly Olney, PhD"
date: "02/17/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
# Ecoli pigs

This workflow uses Seurat v5
Layers in the Seurat v5 object: Seurat v5 assays store data in layers. These layers can store raw, un-normalized counts (layer='counts'), normalized data (layer='data'), or z-scored/variance-stabilized data (layer='scale.data'). 

# Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

# Libraris, paths, colors
```{r echo=FALSE, message=FALSE}
source(here::here("/tgen_labs/jfryer/kolney/Ecoli_pigs/bulk_RNAseq/scripts/", "file_paths_and_colours.R"))
projectID <- "pigs_cellbender"

# Remove sample
# metadata <- subset(metadata, sampleID != "P_60"  & sampleID != "B_6")
metadata$sample_ID <- factor(metadata$sample_ID, levels = c(metadata$sample_ID))
samples <- metadata$sample_ID 
disease_order <- factor(metadata$group, levels = c("Ecoli", "Saline"))
sample_order <- factor(metadata$sample, levels = c(metadata$sample))
```

# Genes 
```{r gene_info}
if (file.exists("../rObjects/annotation.rds")) {
  genes <- readRDS("../rObjects/annotation.rds")
} else {
  gtf.file <- paste0(pathToRef, "Sus_scrofa_star_Ymask_Cellranger_2024/genes/genes.gtf")
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../rObjects/annotation.rds")
}
gene.names <- genes$gene_name
mt.genes.df <- subset(genes, seqnames == "MT")
mt.genes <- mt.genes.df$gene_name
mt.genes <- na.omit(mt.genes)
rm(mt.genes.df)
```

# QC thresholds
nCount_RNA = total number of transcripts (UMIs) in a single cell 
nFeature_RNA = number of unique genes (features)
```{r QC_metrics}
nCount.min <- 500 
nCount.max <- 150000  # 245982 
nFeature.min <- 100
nFeature.max<- 12000
complexity.cutoff <- 0.80
mt.cutoff <- 1
hb.cutoff <- 0
ribo.cutoff <- 2.5
choroid.cutoff <- 1
```

# Setup seurat object
```{r seurat_object}
prefix <- "../cellbender/"
suffix <- "_filtered_seurat.h5"

if (file.exists(paste0("../rObjects/", projectID, ".rds"))) {
  dataObject <- readRDS(paste0("../rObjects/", projectID, ".rds"))
} else {
  for (i in samples) {
    print(i)
    obj <- CreateSeuratObject(Read10X_h5(paste0(prefix,i,"/",i,suffix)))
    assign(i, obj)
  }
  # merge objects
  dataObject <- merge(x = E1_1_BR,
                 y = c(E4_1_BR, E6_1_BR, E8_1_BR, S1_1_BR, S2_1_BR, S3_1_BR, S6_1_BR),
                 add.cell.ids = samples, 
                 project = "Pigs")
    saveRDS(dataObject, paste0("../rObjects/", projectID, ".rds"))
} 
# Inspect 
dataObject

rm(E1_1_BR, E4_1_BR, E6_1_BR, E8_1_BR, S1_1_BR, S2_1_BR, S3_1_BR, S6_1_BR)
```

# Change indents
```{r change_idents}
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
table(sample)
unique(sample)
dataObject$sample <- factor(sample) #, levels = samples)
table(dataObject$sample)  # check
Idents(dataObject) <- dataObject$sample
```

Add condition column to metadata
```{r condition_column}
# Match the samples to the dataObject
sample_match <- match(dataObject$sample, samples)

# Type
dataObject$group <- disease_order[sample_match]
table(dataObject$group) 

dataObject$Sample_ID <- sample_order[sample_match]
table(dataObject$Sample_ID) 
```

QC columns 
```{r QC_columns}
summary(dataObject$nCount_RNA)
summary(dataObject$nFeature_RNA)
# cell.complexity
dataObject$cell.complexity <- log10(dataObject$nFeature_RNA) / log10(dataObject$nCount_RNA)

# Chromosome M
gene.names <- rownames(dataObject)
dataObject$percent.mt <- PercentageFeatureSet(dataObject, features = mt.genes)
summary(dataObject$percent.mt)

# ribosomal proteins 
ribo.genes <- gene.names[grep("^RP[SL]", gene.names)] 
mt.ribo <- gene.names[grep("^MRP[SL]", gene.names)]
ribo.combined <- c(mt.ribo,ribo.genes)
dataObject$percent.ribo <- PercentageFeatureSet(dataObject, features = ribo.combined)
summary(dataObject$percent.ribo)

# hemoglobin proteins
hb.genes <- gene.names[grep("^HB[BA]", gene.names)]
dataObject$percent.hb <- PercentageFeatureSet(dataObject, features = hb.genes)
summary(dataObject$percent.hb)

# percent choroid plexus
dataObject$percent.choroid <- PercentageFeatureSet(dataObject, features = c("TTR","FOLR1", "PRLR"))
summary(dataObject$percent.choroid)
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject$Sample_ID))
colnames(data) <- c("Sample_ID","frequency")

ncells1 <- ggplot(data, aes(x = Sample_ID, y = frequency, fill = Sample_ID)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            #vjust=-0.25, 
            hjust = -.025,
            angle = 90) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,60000, by = 5000), limits = c(0,60000)) +
  ggtitle("Raw: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1
```

## Density plots
```{r prefiltering_density}
# set graphical parameter
par(mfrow = c(4,1))
# Visualize nCount_RNA
den1 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nCount_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) + theme(legend.position="none")


# Visualize nFeature
den2 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = nFeature_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min) +
  geom_vline(xintercept = nFeature.max) + theme(legend.position="none")



# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = cell.complexity,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff) + theme(legend.position="none")

# Visualize percent.mt
den4 <- ggplot(dataObject@meta.data,
       aes(color = Sample_ID,
           x = percent.mt,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff) +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density")+ theme(legend.position="none")



# Arrange graphs in grid
plots1 <- list(den1,den2,den3,den4)
layout1 <- rbind(c(1),c(2),c(3),c(4))
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

## Violin plots
```{r prefiltering_violins}
# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(dataObject,
              features = c( "nCount_RNA","nFeature_RNA", "cell.complexity"),
              ncol = 3,
              group.by = 'Sample_ID',
              ##cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(dataObject,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 4,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v2
```

```{r, echo=FALSE}
# save v1
v1
path <- paste0("../results/violin/",projectID,"_nFeature_nCount_complexity_raw")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 4)
dev.off()

# save v2
v2
path <- paste0("../results/violin/",projectID,"_percent_raw")
saveToPDF(paste0(path, ".pdf"), width = 25, height = 4)
dev.off()

# cleanup
remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter1, warning=FALSE}
s1 <- ggplot(
  dataObject@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_vline(xintercept = nCount.max) + 
  geom_hline(yintercept = nFeature.min) + 
  geom_hline(yintercept = nFeature.max) + 
  facet_wrap(~Sample_ID, ncol = 4) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1
```

```{r echo=FALSE}
# save
s1
path <- paste0("../results/scatter/",projectID,"_nFeature_vs_nCount_percentMt_raw")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 10)
dev.off()

# cleanup
remove(s1)
```
d
# Filtering
## Cell-level filtering
We want to be careful filtering because removing things can easily lead to
misinterpretation.  For example, cells with high percent.mt could actually
just be involved in respiratory processes. \
```{r cell_filtering}
# filter
dataObject.filtered <- subset(dataObject,
                        subset = (nCount_RNA > nCount.min)&
                          (nCount_RNA < nCount.max) & 
                          (nFeature_RNA > nFeature.min) & 
                          (nFeature_RNA < nFeature.max) & 
                          (cell.complexity > complexity.cutoff) &
                          (percent.mt <= mt.cutoff) & 
                          (percent.hb <= hb.cutoff) &
                          (percent.ribo <= ribo.cutoff) &
                          (percent.choroid<= choroid.cutoff))

# print nuclei removed
print(paste0(dim(dataObject)[2] - dim(dataObject.filtered)[2]," nuclei removed"))

table(dataObject$Sample_ID)
table(dataObject.filtered$Sample_ID)
```

## Gene-level filtering
Remove lowly expressed genes.  We will keep genes that have at least 1 count in 10 cells.
```{r gene_filtering}
dataObject.filtered <- JoinLayers(dataObject.filtered)
counts <- GetAssayData(object = dataObject.filtered, layer = "counts")
nonzero <- counts > 0 
keep <- Matrix::rowSums(nonzero) >= 10  
counts.filtered <- counts[keep,]  

dataObject.filtered <- CreateSeuratObject(counts.filtered, 
                                     meta.data = dataObject.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```

## Mitochondrial gene filtering
```{r}
# remove mt.genes
counts <- GetAssayData(object = dataObject.filtered, layer = "counts")
keep <- !rownames(counts) %in% mt.genes # false when mt.gene
counts.filtered <- counts[keep,]

# overwrite dataObject.filtered
dataObject.filtered <- CreateSeuratObject(counts.filtered, 
                                    meta.data = dataObject.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```

## Save filtering
```{r,echo=FALSE}
saveRDS(dataObject.filtered, paste0("../rObjects/",projectID,"_filtered.rds"))
#dataObject.filtered <- readRDS(paste0("../rObjects/",projectID,"_filtered.rds"))
```


## Number of cells
```{r number_cells2}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject.filtered$Sample_ID))
colnames(data) <- c("Sample_ID","frequency")

ncells2 <- ggplot(data, aes(x = Sample_ID, y = frequency, fill = Sample_ID)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            #vjust=-0.25, 
            hjust = -.025,
            angle = 90) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,60000, by = 5000), limits = c(0,60000)) +
  ggtitle("Filtered: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells2
# Arrange graphs in grid
plots2 <- list(ncells1,ncells2)
layout2 <- cbind(c(1),c(2))
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)
```

```{r,echo=FALSE}
# save
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)
path <- paste0("../results/nuclei_count/",projectID, 
               "_cells_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 6)

# cleanup
remove(ncells1,ncells2,plots2,layout2,grid2)
```

## Density plots
```{r postfiltering_density}
# set graphical parameter
par(mfrow = c(4,1))

# Visualize the number of counts per cell
den5 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = nCount_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) + theme(legend.position="none")


den6 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = nFeature_RNA,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min) +
  geom_vline(xintercept = nFeature.max) + theme(legend.position="none")



# Visualize cell complexity
# Quality cells are usually above 0.80
den7 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = cell.complexity,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff) + theme(legend.position="none")


# Visualize percent.mt
den8 <- ggplot(dataObject.filtered@meta.data,
       aes(color = Sample_ID,
           x = percent.mt,
           fill = Sample_ID)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  geom_vline(xintercept = mt.cutoff) + theme(legend.position="none")



# Arrange graphs in grid
plots3 <- list(den1,den2,den3,den4,den5,den6,den7,den8)
layout3 <- rbind(c(1,5),c(2,6),c(3,7), c(4,8))
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
```

```{r, echo=FALSE}
# save
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
path <- paste0("../results/density/",projectID, 
               "_density")
saveToPDF(paste0(path, ".pdf"), width = 11, height = 8.5)

# cleanup
remove(den1,den2,den3,den4,den5,den6,plots3,layout3,grid3)
```

## Violin plots
```{r postfiltering_violins}
# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(dataObject.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(dataObject.filtered,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 4,
              group.by = 'Sample_ID',
              #cols = sample_colors,
              pt.size = 0)
v4
```

```{r, echo=FALSE}
# save
v3
path <- paste0("../results/violin/",projectID, 
               "_nFeature_nCount_complexity_filtered")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)

v4
path <- paste0("../results/violin/",projectID, 
               "_percent_filtered")
saveToPDF(paste0(path, ".pdf"), width = 15, height = 4)

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter1}
s3 <- ggplot(
  dataObject.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_vline(xintercept = nCount.max) + 
  geom_hline(yintercept = nFeature.min) + 
  geom_hline(yintercept = nFeature.max) + 
  facet_wrap(~Sample_ID, ncol = 4) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3
```

```{r,echo=FALSE}
# save
s3
path <- paste0("../results/scatter/",projectID,
               "_nFeature_vs_nCount_perecentMt_filtered")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 10)

# cleanup
remove(s3)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(dataObject.filtered@meta.data,
       aes(x = Sample_ID, 
           y = log10(nFeature_RNA), 
           fill=Sample_ID)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        legend.position = "none") +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Nuclei / Sample") +
  #scale_color_manual(values = sample_colors) +
  #scale_fill_manual(values = sample_colors) +
  xlab("Sample")
b1
```

```{r,echo=FALSE}
# save
b1
path <- paste0("../results/nuclei_count/",projectID,
               "_nFeature_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)

# cleanup
remove(b1)
```

# Top transcripts
```{r top_transcripts}
df <- data.frame(row.names = rownames(dataObject.filtered))
df$rsum <- rowSums(x = dataObject.filtered, slot = "counts")
df$gene_name <- rownames(df)
df <- df[order(df$rsum,decreasing = TRUE),]
head(df, 10)
```

```{r,echo=FALSE}
write.table(df, paste0("../results/top_transcripts/", projectID, "_abundant_transcripts.txt"),
            quote = FALSE,
            row.names = FALSE)
```

```{r}
sessionInfo()
```