---
title: "Manuscript Figures"
author: "Kimberly Olney, PhD"
date: "May 2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# PNAS figure requirements
https://www.pnas.org/author-center/submitting-your-manuscript#manuscript-formatting-guidelines
Small: approximately 9 cm x 6 cm
Medium: approximately 11 cm x 11 cm
Large: approximately 18 cm x 22 cm

Small: approximately 3.54 inches x 2.36 inches
Medium: approximately 4.33 inches x 4.33 inches
Large: approximately 7.09 inches x 8.66 inches


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/tgen_labs/jfryer/kolney/Ecoli_pigs/snRNAseq/scripts/")
setwd("/tgen_labs/jfryer/kolney/Ecoli_pigs/snRNAseq/scripts/")
```

# Libraris, paths, colors
```{r echo=FALSE, message=FALSE}
source(here::here("/tgen_labs/jfryer/kolney/Ecoli_pigs/bulk_RNAseq/scripts/", "file_paths_and_colours.R"))
projectID <- "pigs_cellbender_after_recluster_harm_int_noise_removed_after_annotation_FINAL"
color.panel <- dittoColors()
library(NatParksPalettes)
library("readxl")
library(patchwork) 

my_seurat_theme <- function() {
  theme_classic() + # Start with a base theme
    theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8), 
    theme(legend.position = "none")
    )
}
```

# Read in object
```{r read_object}
# read object
dataObject <- readRDS(paste0("../rObjects/",projectID,"_annotated.rds"))

# inspect
dataObject
Layers(dataObject)
# List of cell types to analyze
cell_types <- c(unique(dataObject$individual_clusters))


group_renaming_map <- c(
  "Ecoli" = "Saline",
  "Saline" = "Ecoli"
)

# 3. Apply the renaming to your specified metadata column
# Make sure to replace 'your_group_column_name' with the actual column name.
dataObject$group <- plyr::mapvalues(
  x = dataObject$group,
  from = names(group_renaming_map),
  to = unname(group_renaming_map)
)

# 4. Verify the changes
unique(dataObject$group)
```
# Figure 1 - bulk RNAseq
### a) bulk volcano
```{r bulk_volcano}
# a) bulk volcano
allComparisons <- c("Ecoli_vs_Saline")
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(paste0("../../bulk_RNAseq/results/DEGs/", i,"_q1.00.txt"))
    i <- gsub("_", " ", i)
  color_values <- vector()
  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$adj.P.Val[row] < 0.05) {
      if (group1_vs_group2$logFC [row] > 1) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$logFC[row] < -1) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  num <- subset(data, (adj.P.Val < 0.05 & logFC < -1)  | (adj.P.Val < 0.05 & logFC > 1 ))
  # Identify up- and down-regulated genes
  up_num <- sum(data$logFC > 1 & data$adj.P.Val < 0.05)
  down_num <- sum(data$logFC < -1 & data$adj.P.Val < 0.05)
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, logFC > 1)
    upFold <- upFold[!(upFold$gene_name %in% up10$gene_name),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, logFC < -1)
    downFold <- downFold[!(downFold$gene_name %in% down10$gene_name),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$P.Value[data$adj.P.Val < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-1)
    posFC <- c(1) 
    bulk_vol_plot <-
      ggplot(data = data,
             aes(
               x = logFC,
               y = -log10(P.Value),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = .8) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i, "\nq < 0.05 & |log2FC| > 1")) +
      geom_text_repel(
        data = up10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 20)
      ) +
      geom_text_repel(
        data = upFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 20)
      ) +
      geom_text_repel(
        data = down10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = downFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8)    
  ) +
  scale_x_continuous(limits = c(-7.5, 10), breaks = seq(-5, 10, 5)) +
  scale_y_continuous(limits = c(0, 12), breaks = seq(0, 12, 2)) +
  annotate("text",
           x = 9.5, # Adjust x position as needed
           y = 11.8,    # Adjust y position as needed
           label = paste("Up =", up_num),
           color = "maroon",
           size = 2.5,
            fontface = "bold",
           hjust = 1, vjust = 1) +
  annotate("text",
           x = -7, # Adjust x position as needed
           y = 11.8,    # Adjust y position as needed
           label = paste("Down =", down_num),
           color = "navyblue",
           fontface = "bold",
           size = 2.5,
           hjust = 0, vjust = 1)
bulk_vol_plot 
    
  }
  } 
```

### b) metascape
```{r bulk_metascape}
# Metascape Enrichment 
#read in enrichment analysis results
up_enrich_results <-
  read_excel(
    paste0(
      "../../bulk_RNAseq/results/metascape/bulk_up/metascape_result.xlsx"
    ), sheet = 2
  )

down_enrich_results <-
  read_excel(
    paste0(
      "../../bulk_RNAseq/results/metascape/bulk_down/metascape_result.xlsx"
    ), sheet = 2
  )

# select the GO term IDs we want to show in the plot
GO_ID_up <-
  c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary",
    "11_Summary",
    "12_Summary",
    "13_Summary",
    "14_Summary",
    "15_Summary",
    "16_Summary",
    "17_Summary",
    "18_Summary",
    "19_Summary", 
    "20_Summary"
  )
GO_ID_down <- c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary",
    "11_Summary",
    "12_Summary",
    "13_Summary",
    "14_Summary",
    "15_Summary",
    "16_Summary",
    "17_Summary",
    "18_Summary",
    "19_Summary", 
    "20_Summary"
)

up_enrich_results_subset <-
  up_enrich_results[up_enrich_results$GroupID %in% GO_ID_up, ]
up_enrich_results_subset$Cluster <- c("upregulated")
up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description,
         levels = up_enrich_results_subset$Description)
up_enrich_results_subset$Description <-
  fct_rev(up_enrich_results_subset$Description)

down_enrich_results_subset <-
  down_enrich_results[down_enrich_results$GroupID %in% GO_ID_down, ]
down_enrich_results_subset$Cluster <- c("downregulated")
down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description,
         levels = down_enrich_results_subset$Description)
down_enrich_results_subset$Description <-
  fct_rev(down_enrich_results_subset$Description)

# get the number of genes in each summary
up_gene_count <-
  strsplit(as.character(up_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
up_gene_count_df <-
  data.frame(matrix(
    unlist(up_gene_count),
    nrow = length(up_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
up_enrich_results_subset$InTerm <- as.numeric(up_gene_count_df$X1)
up_enrich_results_subset$InList <- as.numeric(up_gene_count_df$X2)

down_gene_count <-
  strsplit(as.character(down_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
down_gene_count_df <-
  data.frame(matrix(
    unlist(down_gene_count),
    nrow = length(down_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
down_enrich_results_subset$InTerm <- as.numeric(down_gene_count_df$X1)
down_enrich_results_subset$InList <- as.numeric(down_gene_count_df$X2)

# combine together
up_and_down_enrich_results_subset <- rbind(up_enrich_results_subset, down_enrich_results_subset)

 up_and_down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of response to external stimulus",
         "pos reg of response to external stimulus",
         up_and_down_enrich_results_subset$Description
       )
up_and_down_enrich_results_subset$Description <-
  factor(up_and_down_enrich_results_subset$Description, levels = unique(up_and_down_enrich_results_subset$Description)) 

up_and_down_enrich_results_subset$Description <- fct_rev(up_and_down_enrich_results_subset$Description)

remove(
  down_enrich_results,
  down_gene_count,
  down_gene_count_df,
  up_enrich_results,
  up_gene_count,
  up_gene_count_df
)

up_and_down_enrich_plot <-
  ggplot(data = up_and_down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-45,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_and_down_enrich_plot

addSmallLegend <- function(myPlot, pointSize = 3, textSize = 6, spaceLegend = .5) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
}

# Apply on original plot
up_and_down_enrich_plot <- addSmallLegend(up_and_down_enrich_plot)
```


### Figure 1 combined
```{r figure1_combined}
# Option 2
figure1 <-
  ggarrange(
    bulk_vol_plot,
    up_and_down_enrich_plot,
    NULL, 
    NULL,
    ncol = 2,
    nrow = 2,
  widths = c(2.25, 4.75), 
  labels = c("A","B", "", ""), 
  font.label = list(size = 10)
    )
figure1
path <- paste0("../manuscript_figures/Figure_1_bulk")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 7)
```

# Figure 2 - UMAP & marker genes
a) UMAP, b) relative abundance,  c) bubble plot,
### a) UMAP
```{r UMAP}
# a) UMAP
UMAP_ind <- dittoDimPlot(object = dataObject,
                         var = "individual_clusters",
                         size = .25,
                         reduction.use = "umap",
                         do.label = FALSE,
                         labels.size = 2.5,
                         shape.legend.size = 2,
                         labels.highlight = TRUE,
                         labels.repel = TRUE,
                         legend.show = TRUE)
UMAP_ind <- UMAP_ind +
  ggtitle("Annotated clusters") +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8)    
  )
UMAP_ind
```
### b) relative abundance by group
```{r relative_abundance}
# b) qc
relative_abundance <- dataObject@meta.data %>%
  group_by(individual_clusters, group) %>%
  dplyr::count() %>%
  group_by(group) %>%
  dplyr::mutate(percent = 100 * n / sum(n)) %>%
  ungroup()

rel_abun <- ggplot(relative_abundance, aes(x = group, y = percent, fill = individual_clusters)) +
  geom_col() +
  geom_text(aes(label = paste0(round(percent), "%")), 
            position = position_stack(vjust = 0.5), size = 2.5, color = "white") +
  scale_fill_manual(values = color.panel, name = "Annotated clusters") +
  ggtitle("Percentage of cell type") +
  theme_bw() +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 8),
    plot.title = element_text(size = 8),
    legend.key.height = unit(0.25, "cm"), 
    legend.key.width = unit(0.25, "cm")  
  )
rel_abun
```

### c) bubble plot 
```{r bubble_plot}
# b) bubble plot
desired.cluster.order <- c("oligodendrocyte", "microglia", "astrocyte", "endothelial", 
                           "neuron", "opc", "VLMC", "mural") # order
cluster_rev_order <- rev(desired.cluster.order)
# Convert the identities to a factor with the desired levels
dataObject$individual_clusters <- factor(dataObject$individual_clusters, levels = rev(desired.cluster.order))
Idents(dataObject) <- factor(Idents(dataObject), levels = cluster_rev_order)

markers.to.plot <-
  c(
"CLU", 
"GFAP", 
"AQP4", 
"CLDN5",
"ADGRF5",
"FLT1",
"COL1A1",
"COL1A2",
"C1QA",
"C1QB",
"C1QC",
"ITGAM",
"TYROBP",
"P2RY12",
"RBFOX1",
"SNAP25",
"SYT1",
"GAD1",
"GAD2",
"PLP1",
"MBP", 
"MOG", 
"OLIG1",
"PDGFRA",
"TNR",
"VCAN"
  )

dot_ind <- DotPlot(dataObject,
                   features = markers.to.plot, 
                   idents = cluster_rev_order,
                   dot.scale = 8) + RotatedAxis()
dot_ind <- dot_ind +
  ggtitle("Cluster markers") +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 0),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8, face = "plain")    
  )
dot_ind
```

### Figure 2 combined 
```{r}
# Option 1 with relative abundance plot 
# row1 <- ggarrange(
#   UMAP_ind,
#   rel_abun,
#   ncol = 2,
#   labels = c("a)","b)"), 
#   font.label = list(size = 10)
# )
# row2 <- ggarrange(
#   dot_ind,
#   labels = c("c)"), 
#   font.label = list(size = 10)
#   )
# Option1 <-
#   ggarrange(
#     row1,
#     row2,
#     nrow = 2
#     )
# Option1
# path <- paste0("../manuscript_figures/Figure_2_umap_rel_abun_dotplot")
# saveToPDF(paste0(path, ".pdf"), width = 7, height = 6)

# Option 2
Option1 <-
  ggarrange(
    UMAP_ind,
    dot_ind,
    nrow = 2,
  labels = c("A","B"), 
  font.label = list(size = 10)
    )
Option1
path <- paste0("../manuscript_figures/Figure_2_umap_dotplot")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 6)
```

# Figure 3 - Endothelial 
```{r umap}

endo_umap <- DimPlot(object = subset(dataObject, individual_clusters == "endothelial"),
        reduction = "umap", 
        label = TRUE,
        label.box = TRUE,
        label.size = 3,
        repel = TRUE,
        cols = color.panel[4])

# Create your UMAP plots and apply the custom theme
endo_group_umap <- DimPlot(object = subset(dataObject, individual_clusters == "endothelial"), 
                 reduction = "umap", 
                 cols = c("green3", "gray20"),
                 group.by = "group") +
                 ggtitle("Ecoli") +
                 my_seurat_theme() # Apply your custom theme


endo_HSPH1 <- FeaturePlot(object = subset(dataObject, individual_clusters == "endothelial"),
              reduction = "umap",
              order = TRUE,
              split.by = "group",
              features = c("HSPH1")) +
                 my_seurat_theme()

endo_HSPH1_theme <- endo_HSPH1 & my_seurat_theme() 

```


### a)  volcano
```{r endo_volcano}
# a) endo volcano
allComparisons <- c("endothelial")
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(paste0("../results/DEGs/DESeq2_pseudobulk_exp_filter/", i,"_Ecoli_vs_Saline_comparison_pseudobulk.txt"))
    i <- gsub("_", " ", i)
  color_values <- vector()
    group1_vs_group2 <- na.omit(group1_vs_group2)

  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$padj[row] < 0.05) {
      if (group1_vs_group2$log2FoldChange [row] > 1) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$log2FoldChange[row] < -1) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  num <- subset(data, (padj < 0.05 & log2FoldChange < -1)  | (padj < 0.05 & log2FoldChange > 1 ))
  # Identify up- and down-regulated genes
  up_num <- sum(data$log2FoldChange > 1 & data$padj < 0.05)
  down_num <- sum(data$log2FoldChange < -1 & data$padj < 0.05)
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, log2FoldChange > 5)
    upFold <- upFold[!(upFold$gene %in% up10$gene),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, log2FoldChange < -5)
    downFold <- downFold[!(downFold$gene %in% down10$gene),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$pvalue[data$padj < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-1)
    posFC <- c(1) 
    endo_vol_plot <-
      ggplot(data = data,
             aes(
               x = log2FoldChange,
               y = -log10(pvalue),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = .8) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i, "\nq < 0.05 & |log2FC| > 1")) +
      geom_text_repel(
        data = up10,
        aes(
          x = log2FoldChange,
          y = -log10(pvalue),
          label = gene
        ),
        color = "maroon",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 2)
      ) +
      geom_text_repel(
        data = upFold,
        aes(
          x = log2FoldChange,
          y = -log10(pvalue),
          label = gene
        ),
        color = "maroon",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 2)
      ) +
      geom_text_repel(
        data = down10,
        aes(
          x = log2FoldChange,
          y = -log10(pvalue),
          label = gene
        ),
        color = "navyblue",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 2)
      ) +
      geom_text_repel(
        data = downFold,
        aes(
          x = log2FoldChange,
          y = -log10(pvalue),
          label = gene
        ),
        color = "navyblue",
        fontface = "italic",
        size = 1.8,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 2)
      ) +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8)    
  ) +
  scale_x_continuous(limits = c(-7.5, 12.5), breaks = seq(-5, 12.5, 5)) +
  scale_y_continuous(limits = c(0, 37), breaks = seq(0, 37, 5)) +
  annotate("text",
           x = 6.5, # Adjust x position as needed
           y = 37,    # Adjust y position as needed
           label = paste("Up\n", up_num),
           color = "maroon",
           size = 2.5,
            fontface = "bold",
           hjust = 1, vjust = 1) +
  annotate("text",
           x = -7.5, # Adjust x position as needed
           y = 37,    # Adjust y position as needed
           label = paste("Down\n", down_num),
           color = "navyblue",
           fontface = "bold",
           size = 2.5,
           hjust = 0, vjust = 1)
endo_vol_plot 
    
  }
} 

endo_vol_plot
```

### b) metascape
```{r bulk_metascape}
# Metascape Enrichment 
#read in enrichment analysis results
up_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/endothelial_up/metascape_result.xlsx"
    ), sheet = 2
  )

down_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/endothelial_down/metascape_result.xlsx"
    ), sheet = 2
  )

# select the GO term IDs we want to show in the plot
GO_ID_up <-
  c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

  )
GO_ID_down <- c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

)

up_enrich_results_subset <-
  up_enrich_results[up_enrich_results$GroupID %in% GO_ID_up, ]
up_enrich_results_subset$Cluster <- c("upregulated")
up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description,
         levels = up_enrich_results_subset$Description)
up_enrich_results_subset$Description <-
  fct_rev(up_enrich_results_subset$Description)

down_enrich_results_subset <-
  down_enrich_results[down_enrich_results$GroupID %in% GO_ID_down, ]
down_enrich_results_subset$Cluster <- c("downregulated")
down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description,
         levels = down_enrich_results_subset$Description)
down_enrich_results_subset$Description <-
  fct_rev(down_enrich_results_subset$Description)

# get the number of genes in each summary
up_gene_count <-
  strsplit(as.character(up_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
up_gene_count_df <-
  data.frame(matrix(
    unlist(up_gene_count),
    nrow = length(up_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
up_enrich_results_subset$InTerm <- as.numeric(up_gene_count_df$X1)
up_enrich_results_subset$InList <- as.numeric(up_gene_count_df$X2)

 up_enrich_results_subset$Description <-
       gsub(
         "positive regulation of response to external stimulus",
         "pos reg of response to external stimulus",
         up_enrich_results_subset$Description
       )

 
 up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description, levels = unique(up_enrich_results_subset$Description)) 

up_enrich_results_subset$Description <- fct_rev(up_enrich_results_subset$Description)

 
down_gene_count <-
  strsplit(as.character(down_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
down_gene_count_df <-
  data.frame(matrix(
    unlist(down_gene_count),
    nrow = length(down_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
down_enrich_results_subset$InTerm <- as.numeric(down_gene_count_df$X1)
down_enrich_results_subset$InList <- as.numeric(down_gene_count_df$X2)

down_enrich_results_subset$Description <-
       gsub(
         "plasma membrane bounded cell projection assembly",
         "plasma membrane cell projection",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome segregation",
         "pos reg of chromosome segregation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome organization",
         "pos reg of chromosome organization",
         down_enrich_results_subset$Description
       )



down_enrich_results_subset$Description <-
       gsub(
         "Mitochondrial Fatty Acid Beta-Oxidation",
         "Mito Fatty Acid Beta Oxidation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description, levels = unique(down_enrich_results_subset$Description)) 

down_enrich_results_subset$Description <- fct_rev(down_enrich_results_subset$Description)


# combine together
up_and_down_enrich_results_subset <- rbind(up_enrich_results_subset, down_enrich_results_subset)

up_and_down_enrich_results_subset$Description <-
  factor(up_and_down_enrich_results_subset$Description, levels = unique(up_and_down_enrich_results_subset$Description)) 

up_and_down_enrich_results_subset$Description <- fct_rev(up_and_down_enrich_results_subset$Description)

up_and_down_enrich_plot <-
  ggplot(data = up_and_down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_and_down_enrich_plot

addSmallLegend <- function(myPlot, pointSize = 3, textSize = 6, spaceLegend = .5) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
}

# Apply on original plot
up_and_down_enrich_plot <- addSmallLegend(up_and_down_enrich_plot)

up_enrich_plot <-
  ggplot(data = up_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_enrich_plot

# Apply on original plot
up_enrich_plot <- addSmallLegend(up_enrich_plot)



down_enrich_plot <-
  ggplot(data = down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
down_enrich_plot


# Apply on original plot
down_enrich_plot <- addSmallLegend(down_enrich_plot)
```

### Figure 3 combined
```{r figure1_combined}
row1 <-
  ggarrange(
    endo_vol_plot,
    up_and_down_enrich_plot,
    ncol = 2,
  widths = c(2.25, 4.75), 
  labels = c("A","B"), 
  font.label = list(size = 10)
    )

row2 <-
  ggarrange(
    NULL,
  labels = c("C"), 
  font.label = list(size = 10)
    )

figure3 <-
  ggarrange(
    row1,
    row2,
    nrow = 2,
    heights = c(1, 1.5)
    )
figure3
path <- paste0("../manuscript_figures/Figure_3_endothelial")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 8)
```



# Figure 4 - mural 
### a)  volcano
```{r mural_volcano}
# a) mural volcano
allComparisons <- c("mural")
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(paste0("../results/DEGs/DESeq2_pseudobulk_exp_filter/", i,"_Ecoli_vs_Saline_comparison_pseudobulk.txt"))
    i <- gsub("_", " ", i)
  color_values <- vector()
    group1_vs_group2 <- na.omit(group1_vs_group2)

  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$padj[row] < 0.05) {
      if (group1_vs_group2$log2FoldChange [row] > 1) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$log2FoldChange[row] < -1) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  num <- subset(data, (padj < 0.05 & log2FoldChange < -1)  | (padj < 0.05 & log2FoldChange > 1 ))
  # Identify up- and down-regulated genes
  up_num <- sum(data$log2FoldChange > 1 & data$padj < 0.05)
  down_num <- sum(data$log2FoldChange < -1 & data$padj < 0.05)
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, log2FoldChange > 7)
    upFold <- upFold[!(upFold$gene %in% up10$gene),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, log2FoldChange < -5)
    downFold <- downFold[!(downFold$gene %in% down10$gene),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$pvalue[data$padj < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-1)
    posFC <- c(1) 
    mural_vol_plot <-
      ggplot(data = data,
             aes(
               x = log2FoldChange,
               y = -log10(pvalue),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = .8) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i, "\nq < 0.05 & |log2FC| > 1")) +
      # geom_text_repel(
      #   data = up10,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "maroon",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 20)
      # ) +
      # geom_text_repel(
      #   data = upFold,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "maroon",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
      # geom_text_repel(
      #   data = down10,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "navyblue",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
      # geom_text_repel(
      #   data = downFold,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "navyblue",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8)    
  ) +
  scale_x_continuous(limits = c(-7.5, 12.5), breaks = seq(-5, 12.5, 5)) +
  scale_y_continuous(limits = c(0, 20), breaks = seq(0, 20, 5)) +
  annotate("text",
           x = 6.5, # Adjust x position as needed
           y = 20,    # Adjust y position as needed
           label = paste("Up\n", up_num),
           color = "maroon",
           size = 2.5,
            fontface = "bold",
           hjust = 1, vjust = 1) +
  annotate("text",
           x = -7.5, # Adjust x position as needed
           y = 20,    # Adjust y position as needed
           label = paste("Down\n", down_num),
           color = "navyblue",
           fontface = "bold",
           size = 2.5,
           hjust = 0, vjust = 1)
mural_vol_plot 
    
  }
} 

mural_vol_plot
```

### b) metascape
```{r bulk_metascape}
# Metascape Enrichment 
#read in enrichment analysis results
up_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/mural_up/metascape_result.xlsx"
    ), sheet = 2
  )

down_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/mural_down/metascape_result.xlsx"
    ), sheet = 2
  )

# select the GO term IDs we want to show in the plot
GO_ID_up <-
  c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

  )
GO_ID_down <- c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

)

up_enrich_results_subset <-
  up_enrich_results[up_enrich_results$GroupID %in% GO_ID_up, ]
up_enrich_results_subset$Cluster <- c("upregulated")
up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description,
         levels = up_enrich_results_subset$Description)
up_enrich_results_subset$Description <-
  fct_rev(up_enrich_results_subset$Description)

down_enrich_results_subset <-
  down_enrich_results[down_enrich_results$GroupID %in% GO_ID_down, ]
down_enrich_results_subset$Cluster <- c("downregulated")
down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description,
         levels = down_enrich_results_subset$Description)
down_enrich_results_subset$Description <-
  fct_rev(down_enrich_results_subset$Description)

# get the number of genes in each summary
up_gene_count <-
  strsplit(as.character(up_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
up_gene_count_df <-
  data.frame(matrix(
    unlist(up_gene_count),
    nrow = length(up_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
up_enrich_results_subset$InTerm <- as.numeric(up_gene_count_df$X1)
up_enrich_results_subset$InList <- as.numeric(up_gene_count_df$X2)

 up_enrich_results_subset$Description <-
       gsub(
         "positive regulation of response to external stimulus",
         "pos reg of response to external stimulus",
         up_enrich_results_subset$Description
       )

 
 up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description, levels = unique(up_enrich_results_subset$Description)) 

up_enrich_results_subset$Description <- fct_rev(up_enrich_results_subset$Description)

 
down_gene_count <-
  strsplit(as.character(down_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
down_gene_count_df <-
  data.frame(matrix(
    unlist(down_gene_count),
    nrow = length(down_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
down_enrich_results_subset$InTerm <- as.numeric(down_gene_count_df$X1)
down_enrich_results_subset$InList <- as.numeric(down_gene_count_df$X2)

down_enrich_results_subset$Description <-
       gsub(
         "plasma membrane bounded cell projection assembly",
         "plasma membrane cell projection",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome segregation",
         "pos reg of chromosome segregation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome organization",
         "pos reg of chromosome organization",
         down_enrich_results_subset$Description
       )



down_enrich_results_subset$Description <-
       gsub(
         "Mitochondrial Fatty Acid Beta-Oxidation",
         "Mito Fatty Acid Beta Oxidation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description, levels = unique(down_enrich_results_subset$Description)) 

down_enrich_results_subset$Description <- fct_rev(down_enrich_results_subset$Description)


# combine together
up_and_down_enrich_results_subset <- rbind(up_enrich_results_subset, down_enrich_results_subset)

up_and_down_enrich_results_subset$Description <-
  factor(up_and_down_enrich_results_subset$Description, levels = unique(up_and_down_enrich_results_subset$Description)) 

up_and_down_enrich_results_subset$Description <- fct_rev(up_and_down_enrich_results_subset$Description)

up_and_down_enrich_plot <-
  ggplot(data = up_and_down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_and_down_enrich_plot

addSmallLegend <- function(myPlot, pointSize = 3, textSize = 6, spaceLegend = .5) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
}

# Apply on original plot
up_and_down_enrich_plot <- addSmallLegend(up_and_down_enrich_plot)

up_enrich_plot <-
  ggplot(data = up_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_enrich_plot

# Apply on original plot
up_enrich_plot <- addSmallLegend(up_enrich_plot)



down_enrich_plot <-
  ggplot(data = down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
down_enrich_plot


# Apply on original plot
down_enrich_plot <- addSmallLegend(down_enrich_plot)
```

### Figure 4 combined
```{r figure1_combined}
row1 <-
  ggarrange(
    mural_vol_plot,
    up_and_down_enrich_plot,
    ncol = 2,
  widths = c(2.25, 4.75), 
  labels = c("A","B"), 
  font.label = list(size = 10)
    )

row2 <-
  ggarrange(
    NULL,
  labels = c("C"), 
  font.label = list(size = 10)
    )

figure3 <-
  ggarrange(
    row1,
    row2,
    nrow = 2,
    heights = c(1, 1.5)
    )
figure3
path <- paste0("../manuscript_figures/Figure_4_mural")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 8)
```

# Figure 5 - microglia 
### a)  volcano
```{r microglia_volcano}
# a) microglia volcano
allComparisons <- c("microglia")
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(paste0("../results/DEGs/DESeq2_pseudobulk_exp_filter/", i,"_Ecoli_vs_Saline_comparison_pseudobulk.txt"))
    i <- gsub("_", " ", i)
  color_values <- vector()
    group1_vs_group2 <- na.omit(group1_vs_group2)

  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$padj[row] < 0.05) {
      if (group1_vs_group2$log2FoldChange [row] > 1) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$log2FoldChange[row] < -1) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  num <- subset(data, (padj < 0.05 & log2FoldChange < -1)  | (padj < 0.05 & log2FoldChange > 1 ))
  # Identify up- and down-regulated genes
  up_num <- sum(data$log2FoldChange > 1 & data$padj < 0.05)
  down_num <- sum(data$log2FoldChange < -1 & data$padj < 0.05)
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, log2FoldChange > 7)
    upFold <- upFold[!(upFold$gene %in% up10$gene),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, log2FoldChange < -5)
    downFold <- downFold[!(downFold$gene %in% down10$gene),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$pvalue[data$padj < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-1)
    posFC <- c(1) 
    microglia_vol_plot <-
      ggplot(data = data,
             aes(
               x = log2FoldChange,
               y = -log10(pvalue),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = .8) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i, "\nq < 0.05 & |log2FC| > 1")) +
      # geom_text_repel(
      #   data = up10,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "maroon",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 20)
      # ) +
      # geom_text_repel(
      #   data = upFold,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "maroon",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
      # geom_text_repel(
      #   data = down10,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "navyblue",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
      # geom_text_repel(
      #   data = downFold,
      #   aes(
      #     x = log2FoldChange,
      #     y = -log10(pvalue),
      #     label = gene
      #   ),
      #   color = "navyblue",
      #   fontface = "italic",
      #   size = 1.8,
      #   max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      # ) +
  theme(
    axis.title.x = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title.y = element_text(size = 8),
    axis.text.y = element_text(size = 8),
    legend.title = element_text(size = 8), 
    legend.text = element_text(size = 8),  
    plot.title = element_text(size = 8)    
  ) +
  scale_x_continuous(limits = c(-7.5, 12.5), breaks = seq(-5, 12.5, 5)) +
  scale_y_continuous(limits = c(0, 20), breaks = seq(0, 20, 5)) +
  annotate("text",
           x = 6.5, # Adjust x position as needed
           y = 20,    # Adjust y position as needed
           label = paste("Up\n", up_num),
           color = "maroon",
           size = 2.5,
            fontface = "bold",
           hjust = 1, vjust = 1) +
  annotate("text",
           x = -7.5, # Adjust x position as needed
           y = 20,    # Adjust y position as needed
           label = paste("Down\n", down_num),
           color = "navyblue",
           fontface = "bold",
           size = 2.5,
           hjust = 0, vjust = 1)
microglia_vol_plot 
    
  }
} 

microglia_vol_plot
```

### b) metascape
```{r bulk_metascape}
# Metascape Enrichment 
#read in enrichment analysis results
up_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/microglia_up/metascape_result.xlsx"
    ), sheet = 2
  )

down_enrich_results <-
  read_excel(
    paste0(
      "../results/metascape/microglia_down/metascape_result.xlsx"
    ), sheet = 2
  )

# select the GO term IDs we want to show in the plot
GO_ID_up <-
  c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

  )
GO_ID_down <- c(
    "1_Summary",
    "2_Summary",
    "3_Summary",
    "4_Summary",
    "5_Summary",
    "6_Summary",
    "7_Summary",
    "8_Summary",
    "9_Summary", 
    "10_Summary"

)

up_enrich_results_subset <-
  up_enrich_results[up_enrich_results$GroupID %in% GO_ID_up, ]
up_enrich_results_subset$Cluster <- c("upregulated")
up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description,
         levels = up_enrich_results_subset$Description)
up_enrich_results_subset$Description <-
  fct_rev(up_enrich_results_subset$Description)

down_enrich_results_subset <-
  down_enrich_results[down_enrich_results$GroupID %in% GO_ID_down, ]
down_enrich_results_subset$Cluster <- c("downregulated")
down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description,
         levels = down_enrich_results_subset$Description)
down_enrich_results_subset$Description <-
  fct_rev(down_enrich_results_subset$Description)

# get the number of genes in each summary
up_gene_count <-
  strsplit(as.character(up_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
up_gene_count_df <-
  data.frame(matrix(
    unlist(up_gene_count),
    nrow = length(up_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
up_enrich_results_subset$InTerm <- as.numeric(up_gene_count_df$X1)
up_enrich_results_subset$InList <- as.numeric(up_gene_count_df$X2)

 up_enrich_results_subset$Description <-
       gsub(
         "positive regulation of response to external stimulus",
         "pos reg of response to external stimulus",
         up_enrich_results_subset$Description
       )

 
 up_enrich_results_subset$Description <-
  factor(up_enrich_results_subset$Description, levels = unique(up_enrich_results_subset$Description)) 

up_enrich_results_subset$Description <- fct_rev(up_enrich_results_subset$Description)

 
down_gene_count <-
  strsplit(as.character(down_enrich_results_subset$InTerm_InList),
           "/",
           fixed = T)
down_gene_count_df <-
  data.frame(matrix(
    unlist(down_gene_count),
    nrow = length(down_gene_count),
    byrow = TRUE
  ), stringsAsFactors = FALSE)
down_enrich_results_subset$InTerm <- as.numeric(down_gene_count_df$X1)
down_enrich_results_subset$InList <- as.numeric(down_gene_count_df$X2)

down_enrich_results_subset$Description <-
       gsub(
         "plasma membrane bounded cell projection assembly",
         "plasma membrane cell projection",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome segregation",
         "pos reg of chromosome segregation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
       gsub(
         "positive regulation of chromosome organization",
         "pos reg of chromosome organization",
         down_enrich_results_subset$Description
       )



down_enrich_results_subset$Description <-
       gsub(
         "Mitochondrial Fatty Acid Beta-Oxidation",
         "Mito Fatty Acid Beta Oxidation",
         down_enrich_results_subset$Description
       )

down_enrich_results_subset$Description <-
  factor(down_enrich_results_subset$Description, levels = unique(down_enrich_results_subset$Description)) 

down_enrich_results_subset$Description <- fct_rev(down_enrich_results_subset$Description)


# combine together
up_and_down_enrich_results_subset <- rbind(up_enrich_results_subset, down_enrich_results_subset)

up_and_down_enrich_results_subset$Description <-
  factor(up_and_down_enrich_results_subset$Description, levels = unique(up_and_down_enrich_results_subset$Description)) 

up_and_down_enrich_results_subset$Description <- fct_rev(up_and_down_enrich_results_subset$Description)

up_and_down_enrich_plot <-
  ggplot(data = up_and_down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_and_down_enrich_plot

addSmallLegend <- function(myPlot, pointSize = 3, textSize = 6, spaceLegend = .5) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
}

# Apply on original plot
up_and_down_enrich_plot <- addSmallLegend(up_and_down_enrich_plot)

up_enrich_plot <-
  ggplot(data = up_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
up_enrich_plot

# Apply on original plot
up_enrich_plot <- addSmallLegend(up_enrich_plot)



down_enrich_plot <-
  ggplot(data = down_enrich_results_subset, aes(x = InTerm, y = Description)) +
  ggplot2::facet_grid(~ Cluster, scales = "free") +
  geom_bar(stat = "identity", aes(fill = LogP), width = .7, position = position_dodge(width = .2)) +
  theme_bw() +
 # ggtitle(paste0(tissue," enrichment summaries")) +
  labs(x = "Gene count", y = NULL) +
  guides(fill = guide_legend(title = expression(log[10] ~ "(" ~ italic("p") ~ "-value)"))) +
  scale_fill_gradientn(
    colours = c(colors=natparks.pals("Denali")),
    guide = "legend",
    limits = c(-52,-1)
  ) +
    theme(strip.text = element_text(size = 8), 
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8), 
          axis.title.x = element_text(size = 8), 
          legend.text = element_text(size = 8),
          legend.title = element_text(size = 8),
          legend.margin=margin(0,0.5,0,0),
          legend.box.margin=margin(-10,-2,-10,-7.5), 
          plot.margin = margin(0.1, 0.2, 0, 0.2, "cm"), 
          plot.title = element_text(size = 8, hjust = -3.25, vjust=0, margin = margin(0,0,0,0)))
down_enrich_plot


# Apply on original plot
down_enrich_plot <- addSmallLegend(down_enrich_plot)
```

### Figure 5 combined
```{r figure1_combined}
row1 <-
  ggarrange(
    microglia_vol_plot,
    up_and_down_enrich_plot,
    ncol = 2,
  widths = c(2.25, 4.75), 
  labels = c("A","B"), 
  font.label = list(size = 10)
    )

row2 <-
  ggarrange(
    NULL,
  labels = c("C"), 
  font.label = list(size = 10)
    )

figure3 <-
  ggarrange(
    row1,
    row2,
    nrow = 2,
    heights = c(1, 1.5)
    )
figure3
path <- paste0("../manuscript_figures/Figure_5_microglia")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 8)
```





