configfile: "config.json"
def directory(path):
    return path

cellranger_path = "/tgen_labs/jfryer/kolney/tools/cellranger-9.0.0/bin/cellranger"

# rule order
#ruleorder: cellranger > move 

rule all:
    input:
        expand(
            "{fastq_path}{seq_order}_L003_I1_001.fastq.gz",
            fastq_path=config["fastq_path"],
            seq_order=[config[sample]["fq"] for sample in config["sample_names"]]
        ),
        expand(directory("{sample}/outs"), sample=config["sample_names"])
        #expand(config["cellranger_dir"]+"{sample}/", sample=config["sample_names"])


#------------------------------
# make reference 
#------------------------------

# /tgen_labs/jfryer/kolney/tools/cellranger-9.0.0/bin/cellranger mkref --genome=Sus_scrofa_star_Ymask_Cellranger_2024 --fasta=Sus_scrofa.Sscrofa11.1.dna.toplevel.Ymask.fa --genes=Sus_scrofa.Sscrofa11.1.107.gtf

#------------------------------
# alignment
#------------------------------
rule cellranger:
    params:
        cellranger = cellranger_path,
        id = lambda wildcards: config[wildcards.sample]["ID"],
       # sm = lambda wildcards: config[wildcards.sample]["SM"],
        path = lambda wildcards: config[wildcards.sample]["fq_path"],
        ref = config["Sus.scrofa"]
    output:
        "{sample}/outs"
    shell:
        """
        {params.cellranger} count --id={params.id} --sample={params.id} --fastqs={params.path} --transcriptome={params.ref} --create-bam true --localcores=16 --localmem=50 || true
        """

# $sample is the sampleID (e.g. 11-87_BR)
# --fastqs is path to the snRNAseq fastq files
# --transcriptome is the path to the human genome directory. This was created in a prior step.
# --localcores will restrict cellranger to 16 cores
# --localmem will restrict cellranger to 50G memory
